brainiak/matnorm_models/base.py

```diff
@@ -0,0 +1,229 @@
+                      (n+1, current_loss))
+                break
+            if (n+1) % step == 0:
+                print('iter %i, %f' % (n+1, current_loss))
+
+    def solve_det_marginal(self, x, sigma, A, Q):
+        """
+        Use matrix inversion lemma for the solve:
+        .. math::
+            (\Sigma + AQA')^{-1} X =\\
+             \Sigma^{-1} - \Sigma^{-1} A (Q^{-1} + A' \Sigma^{-1} A)^{-1} A' \Sigma^{-1}
+
+        Use matrix determinant lemma for determinant:
+        ..math::
+            \log|(\Sigma + AQA')| = \log|Q^{-1} + A' \Sigma^{-1} A| + \log|Q| + \log|\Sigma|
+        """
+
+        # cholesky of (Qinv + A' Sigma^{-1} A)
+        i_qf_cholesky = tf.cholesky(Q.Sigma_inv + tf.matmul(A,
+                                    sigma.Sigma_inv_x(A), transpose_a=True))
+
+        logdet = Q.logdet + sigma.logdet +\
+            2 * tf.reduce_sum(tf.log(tf.matrix_diag_part(i_qf_cholesky)))
+
+        # A' Sigma^{-1}
+        Atrp_Sinv = tf.matmul(A, sigma.Sigma_inv, transpose_a=True)
+        # (Qinv + A' Sigma^{-1} A)^{-1} A' Sigma^{-1}
+        prod_term = tf.cholesky_solve(i_qf_cholesky, Atrp_Sinv)
+        solve = tf.matmul(sigma.Sigma_inv -
+                          sigma.Sigma_inv_x(tf.matmul(A, prod_term)), x)
+        return solve, logdet
+
+    def solve_det_conditional(self, x, sigma, A, Q):
+        """
+        Use matrix inversion lemma for the solve:
+        .. math::
+            (\Sigma - AQA')^{-1} X =\\
+             \Sigma^{-1} + \Sigma^{-1} A (Q^{-1} - A' \Sigma^{-1} A)^{-1} A' \Sigma^{-1}
+
+        Use matrix determinant lemma for determinant:
+        ..math::
+            \log|(\Sigma - AQA')| = \log|Q^{-1} - A' \Sigma^{-1} A| + \log|Q| + \log|\Sigma|
+        """
+
+        # (Qinv - A' Sigma^{-1} A)
+        i_qf_cholesky = tf.cholesky(Q.Sigma - tf.matmul(A,
+                                    sigma.Sigma_inv_x(A), transpose_a=True))
+
+        logdet = -Q.logdet + sigma.logdet +\
+            2 * tf.reduce_sum(tf.log(tf.matrix_diag_part(i_qf_cholesky)))
+
+        # A' Sigma^{-1}
+        Atrp_Sinv = tf.matmul(A, sigma.Sigma_inv, transpose_a=True)
+        # (Q - A' Sigma^{-1} A)^{-1} A' Sigma^{-1}
+        prod_term = tf.cholesky_solve(i_qf_cholesky, Atrp_Sinv)
+
+        solve = tf.matmul(sigma.Sigma_inv +
+                          sigma.Sigma_inv_x(tf.matmul(A, prod_term)), x)
+
+        return solve, logdet
+
+    def _mnorm_logp_internal(self, colsize, rowsize, logdet_row, logdet_col,
+                             solve_row, solve_col):
+        """Construct logp from the solves and determinants.
+        """
+        log2pi = 1.8378770664093453
+
+        denominator = - rowsize * colsize * log2pi -\
+            colsize * logdet_row - rowsize * logdet_col
+        numerator = - tf.trace(tf.matmul(solve_col, solve_row))
+        return tf.reduce_sum(0.5 * (numerator + denominator))
+
+    def matnorm_logp(self, x, row_cov, col_cov):
```

Comment:
In general I would agree. However, the way MatnormModelBase is subclassed makes this less clear from context. For example, in matnom-BRSA we currently call `self.matnorm_logp_marginal_row`. If this instead were `self.logp_marginal_row` it is less clear what it is the log likelihood for exactly (it is not for BRSA, it is specifically for a matrix varaite normal with a marginalization term in the row covariance). 

On reflection I wonder if what this is telling me is that I those likelihoods should live in their own namespace and not be class methods (so instead of `self.matnorm_logp` we would `from matnorm_likelihoods import matnorm_logp` and then just use it. Thoughts?